src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -178,10 +178,46 @@ void Terminal::UpdateSettings(winrt::Microsoft::Terminal::Settings::ICoreSetting
 
     const short newBufferHeight = viewportSize.Y + _scrollbackLines;
     COORD bufferSize{ viewportSize.X, newBufferHeight };
-    RETURN_IF_FAILED(_buffer->ResizeTraditional(bufferSize));
 
-    auto proposedTop = oldTop;
-    const auto newView = Viewport::FromDimensions({ 0, proposedTop }, viewportSize);
+    // First allocate a new text buffer to take the place of the current one.
+    std::unique_ptr<TextBuffer> newTextBuffer;
+    try
+    {
+        newTextBuffer = std::make_unique<TextBuffer>(bufferSize,
+                                                     _buffer->GetCurrentAttributes(),
+                                                     0, // temporarily set size to 0 so it won't render.
+                                                     _buffer->GetRenderTarget());
+    }
+    CATCH_RETURN();
+
+    RETURN_IF_FAILED(TextBuffer::Reflow(*_buffer.get(), *newTextBuffer.get(), _mutableViewport));
+
+    // However conpty resizes a little oddly - if the height decreased, and
+    // there were blank lines at the bottom, those lines will get trimmed.
+    // If there's not blank lines, then the top will get "shifted down",
+    // moving the top line into scrollback.
+    // See GH#3490 for more details.
+
+    // If the final position in the buffer is on the bottom row of the new
+    // viewport, then we're going to need to move the top down. Otherwise,
+    // move the bottom up.
+    const auto dy = viewportSize.Y - oldDimensions.Y;
+    const COORD oldCursorPos = _buffer->GetCursor().GetPosition();
+    COORD oldLastChar = oldCursorPos;
+    try
+    {
+        oldLastChar = _buffer->GetLastNonSpaceCharacter(_mutableViewport);
+    }
+    CATCH_LOG();
+
+    const auto maxRow = std::max(oldLastChar.Y, oldCursorPos.Y);
+
+    const bool beforeLastRowOfView = maxRow < _mutableViewport.BottomInclusive();
+    const auto adjustment = beforeLastRowOfView ? 0 : std::max(0, -dy);
+
+    auto proposedTop = oldTop + adjustment;
+
+    const auto newView = Viewport::FromDimensions({ 0, gsl::narrow_cast<short>(proposedTop) }, viewportSize);

```

Comment:
can we use chromium math here instead of the narrow_cast?

Reply:
Lol I wrote most of this PR before we had that library 😝 Yea I'll update it